src/martialarts.cpp

```diff
@@ -2449,7 +2449,9 @@ void ma_details_ui_impl::draw_ma_details_text() const
         for( const auto &entry : weapons_text ) {
             cataimgui::draw_colored_text( string_format( _( "<header>%s</header>" ), entry.first ) );
```

Comment:
Since you bring that up regularly, what are the gains of doing it like that? It comes at the price that the code becomes more complex and harder to read, essentially duplicating parts of the functionality of `draw_colored_text`. So at least in my opinion that needs a strong justification, but this looks like just premature optimization.

Reply:
It’s not a performance optimization _yet_. Because with ImGui we rerender the UI at a constant rate, allocating new strings every single frame not only wastes time, it also fragments memory. Memory fragmentation is pernicious; it makes every future memory operation slower. It can even cause allocations to fail, even while your memory is half unused.

I would prefer if you considered `cataimgui::draw_colored_text` and `cataimgui::TextColoredParagraph` to be duplicating the existing functionality of ImGui, rather than the other way around. They are there to allow modders and translators to insert colors into static text and art, but if we are building a string containing tags by concatenation or substitution at run time then we are doing it wrong.

Reply:
That does not answer my question what the gains would actually be. Are there any _measurable_ effects from doing it one way or the other?

Reply:
No, not yet. If we write the whole program so that it allocates madly on every frame, then we’ll fragment all of the ram, run out of real memory, and start swapping. Then our poor performance will be really noticeable!

It’s like weeding a garden. One weed won’t do any harm, but if you put off weeding until there are thousands of weeds then you’re really going to regret it.

Reply:
You're still talking purely hypothetical. And even then the solution is not "don't use our custom wrappers".

Reply:
No, I speak from experience. The solution is to use ImGui correctly, and avoid allocating on every frame. 

In this window, for example, the current implementation builds a vector of strings once and then displays them every frame. Building the vector requires allocating (and probably more than once as the vector grows), plus copying the `char*`s from the l18n layer into `std::string`s for each line. A more idiomatic implementation would simply display the strings, and forego the vector. I didn’t object to the vector of strings in my review because it kept the diff smaller, and because the vector is only created once. A more idiomatic implementation would look something like this:

```c++
if( !general_info_text.empty() &&
    ImGui::CollapsingHeader( _( "General info" ),
                             general_info_group_collapsed ? ImGuiTreeNodeFlags_None : ImGuiTreeNodeFlags_DefaultOpen ) ) {
  if( ma.force_unarmed ) {
    cataimgui::TextParagraphColored( _( "This style <bold>forces you to use unarmed strikes</bold>, even if wielding a weapon." ) );
    ImGui::NewLine();
  } else if( ma.allow_all_weapons ) {
    cataimgui::TextParagraphColored( _( "This style can be used with <bold>all</bold> weapons." ) );
    ImGui::NewLine();
  } else if( ma.strictly_melee ) {
    // etc…
  }
  // …
}
```

This does not allocate anything. It doesn’t build up state and then display it. Instead we simply use the state of the `ma` object itself to guide the display. Note that the `_` function returns a `char*`, a pointer to a buffer owned by the translation cache. We don’t copy it into a `std::string`. We don't call `replace_colors` on it, or anything like that. We don't pass it a `std::string` that it would be forced to copy, either, but as far as I am aware that doesn’t happen much in this program (thankfully).

But again, I don’t insist that the code be immediately rewritten in this form. What we have now is good enough, and easier to review.